software/apps/cholesky_q32s/main.c

```diff
@@ -0,0 +1,196 @@
+  init_matrix_zeros(M_matrix, N, N, core_id);
+#if defined(FOLDED)
+  init_matrix_zeros(LL_matrix, N, N_BANKS, core_id);
+  init_matrix_zeros(LR_matrix, N, N_BANKS, core_id);
+#else
+  init_matrix_zeros(L_matrix, N, N, core_id);
+#endif
+  mempool_barrier(num_cores);
+  /* Create positive definite matrix */
+  if (core_id == 0) {
+    transpose(A_matrix, AT_matrix, N);
+    matrixmult(AT_matrix, A_matrix, M_matrix, N);
+  }
+  mempool_barrier(num_cores);
+}
+
+// Driver program
+void single_core_cholesky() {
+  uint32_t core_id = mempool_get_core_id();
+  uint32_t num_cores = mempool_get_core_count();
+  mempool_barrier_init(core_id); // Initialize barrier and synchronize
+  /* Initialize matrices */
+  initialize();
+  /* Benchmark */
+  if (core_id == 0) {
+    mempool_start_benchmark();
+    mempool_cholesky_crout_q32s(M_matrix, L_matrix, N);
+    mempool_stop_benchmark();
+  }
+  mempool_barrier(num_cores);
+}
+
+// Driver program
+void single_core_linearsolver() {
+  uint32_t core_id = mempool_get_core_id();
+  uint32_t num_cores = mempool_get_core_count();
+  mempool_barrier_init(core_id); // Initialize barrier and synchronize
+  /* Initialize matrices */
+  initialize();
+  /* Benchmark */
+  if (core_id == 0) {
+    mempool_start_benchmark();
+    mempool_linearsolver_q32s(M_matrix, L_matrix, In, N);
+    mempool_uprtrisolver_q32s(L_matrix, In, N);
+    mempool_stop_benchmark();
+  }
+  mempool_barrier(num_cores);
+}
+
+// Driver program
+void parallel_cholesky() {
+  uint32_t core_id = mempool_get_core_id();
+  uint32_t num_cores = mempool_get_core_count();
+  mempool_barrier_init(core_id); // Initialize barrier and synchronize
+  /* Initialize matrices */
+  initialize();
+  /* Benchmark */
+  mempool_start_benchmark();
+  mempool_cholesky_q32p(M_matrix, L_matrix, N);
+  mempool_stop_benchmark();
+  mempool_barrier(num_cores);
+}
+
+// Driver program
+void parallel_cholesky_folded() {
+  uint32_t core_id = mempool_get_core_id();
+  uint32_t num_cores = mempool_get_core_count();
+  uint32_t nPE = N / 4;
+  mempool_barrier_init(core_id); // Initialize barrier and synchronize
+  /* Initialize matrices */
+  initialize();
+  /* Benchmark */
+  if (core_id < nPE) {
+    mempool_start_benchmark();
+    mempool_cholesky_q32p_fold(M_matrix, M_matrix, LL_matrix, LR_matrix, N,
+                               nPE);
+    mempool_stop_benchmark();
+  }
+  mempool_barrier(num_cores);
+#ifdef VERBOSE
+  if (core_id == 0) {
+    display(LL_matrix, N, N);
+    display(LR_matrix, N, N);
+    display(In, 1, N);
+  }
+  mempool_barrier(num_cores);
+#endif
+}
+
+// Driver program
+void parallel_linearsolver_folded() {
+  uint32_t core_id = mempool_get_core_id();
+  uint32_t num_cores = mempool_get_core_count();
+  mempool_barrier_init(core_id); // Initialize barrier and synchronize
+  /* Initialize matrices */
+  initialize();
+  /* Benchmark */
+  uint32_t nPE = N / 4;
+  if (core_id < nPE) {
+    mempool_start_benchmark();
```

Comment:
This part is pretty much the same for all those functions, and you have the barrier initialization in the main function as well, right? You could remove a lot of code duplication and indirection if you directly call, for example, ` mempool_linearsolver_q32p_fold(M_matrix, M_matrix, LL_matrix, LR_matrix, In, In, N, nPE);` in the main function instead of wrapping it in the `parallel_linearsolver_folded` function, or am I missing something?

Reply:
Ok, I will call the function from the main, but I still want to keep the redundancy in the loops for calling it, because in my opinion, they make the code more readable